devbin/krsync.sh

```diff
@@ -0,0 +1,22 @@
+#!/bin/bash
+# https://serverfault.com/questions/741670/rsync-files-to-a-kubernetes-pod
+
+if [ -z "$KRSYNC_STARTED" ]; then
+    export KRSYNC_STARTED=true
+    exec rsync --blocking-io --rsh "$0" $@
+fi
+
+# Running as --rsh
+namespace=''
+pod=$1
+shift
+
```

Comment:
Can you explain how this is structured? When would KRSYNC_STARTED ever be set? Also, I thought the argument to `--rsh` was a shell. Finally, how is this block below used and when?

Reply:
The first time this file is invoked, `KRSYNC_STARTED` is not set so we enter the if statement at line 4 and call `rsync`, telling rsync that it should use *this script* as `ssh`.

Rsync then calls *back into* us (sync.py -> krsync.sh -> rsync -> krsync.sh). If rsync sees `/from pod:/to`, it calls us with the arguments `pod ...`.  If rsync sees `/from pod@namespace:/to`, it calls us with the arguments `-l pod namespace`. We optimistically assume the first form on line 11. then we shift away the first argument. If the first argument was actually `-l`, then the if statement on line 15 will succeed. In that case, the second (now first) and third (now second) arguments are the pod and namespace. Finally we call kubectl exec, which is effectively a kubectl version of ssh.

So there's like three levels of masquerading! This script, at first, masquerades as `rsync`. Then rsync calls this script, who is now masquerading as `ssh`. Then this script calls `kubectl exec` who also masquerades as `ssh`.